graylog2-server/src/main/java/org/graylog2/security/ldap/LdapConnector.java

```diff
@@ -99,32 +106,84 @@ public LdapEntry search(LdapNetworkConnection connection, String searchBase, Str
-                                                          filter,
-                                                          SearchScope.SUBTREE,
-                                                          "*");
-        final Iterator<Entry> it = entryCursor.iterator();
-        if (it.hasNext()) {
-            final Entry e = it.next();
-            // for generic LDAP use the dn of the entry for the subsequent bind, active directory needs the userPrincipalName attribute (set below)
-            if (!activeDirectory) {
+        EntryCursor entryCursor = null;
+        try {
+            entryCursor = connection.search(searchBase,
+                                            filter,
+                                            SearchScope.SUBTREE,
+                                            "*");
+            final Iterator<Entry> it = entryCursor.iterator();
+            if (it.hasNext()) {
+                final Entry e = it.next();
+                // always set the proper DN for the entry, we need it for group matching
                 ldapEntry.setDn(e.getDn().getName());
+
+                // for generic LDAP use the dn of the entry for the subsequent bind, active directory needs the userPrincipalName attribute (set below)
+                if (!activeDirectory) {
+                    ldapEntry.setBindPrincipal(e.getDn().getName());
+                }
+
+                for (Attribute attribute : e.getAttributes()) {
+                    if (activeDirectory && attribute.getId().equalsIgnoreCase("userPrincipalName")) {
+                        ldapEntry.setBindPrincipal(attribute.getString());
+                    }
+                    if (attribute.isHumanReadable()) {
+                        ldapEntry.put(attribute.getId(), attribute.getString());
+                    }
+                }
+            } else {
+                LOG.trace("No LDAP entry found for filter {}", filter);
+                return null;
             }
+            // TODO ActiveDirectory could use the memberOf attribute, but then we'd need to resolve the CN of each group, too.
+            // TODO we do not check for dynamic groups yet.
+            ldapEntry.addGroups(findGroups(connection, "CN=Builtin,DC=corp,DC=graylog,DC=org", "group", "cn", ldapEntry.getDn()));
+            LOG.trace("LDAP search found entry for DN {} with search filter {}: {}", ldapEntry.getDn(), filter, ldapEntry);
+            return ldapEntry;
+        } finally {
+            if (entryCursor != null) {
+                entryCursor.close();
+            }
+        }
+    }
+
+    public Set<String> findGroups(LdapNetworkConnection connection,
+                                  String groupSearchBase,
+                                  String groupObjectClass,
+                                  String groupIdAttribute,
+                                  String dn) throws LdapException {
+        final HashSet<String> groups = Sets.newHashSet();
```

Comment:
Minor code-smell: Using a concrete implementation instead of interface at variable declaration.